File: AnotherApproach/pvector.cpp
#include "pvector.h"

Pvector::Pvector(float _x, float _y)
{
    x = _x;
    y = _y;
}

void Pvector::add(Pvector *velo){
    x += velo->x;
    y += velo->y;
}

void Pvector::sub(Pvector *velo){
    x += velo->x;
    y += velo->y;
}

void Pvector::mult(float n){
    this->x *= n;
    this->y *= n;
}

void Pvector::div(float n){
    this->x /= n;
    this->y /= n;
}

float Pvector::mag(){
    return sqrtf(this->x*this->x + this->y*this->y);
}

void Pvector::normalize(){
    float m = mag();
    if(m != 0){
        div(m);
    }
}

Pvector Pvector::get(){
    Pvector *v1 = new Pvector(0,0);
    *v1 = *this;
    return *v1;
}

void Pvector::setMag(float x)
{
    this->normalize();
    this->mult(x);
}


void Pvector::limit(float max){
    if (this->mag() > max)
    {
        this->setMag(max);
    }
}

Pvector* Pvector::add(Pvector* v1,Pvector* v2){
    Pvector* v3 = new Pvector(v1->x + v2->x, v1->y + v2->y);
    return v3;
}

Pvector* Pvector::sub(Pvector* v1,Pvector* v2){
    Pvector* v3 = new Pvector(v1->x - v2->x, v1->y - v2->y);
    return v3;
}

Pvector* Pvector::mult(Pvector* v1, float n){
    Pvector *v3 = new Pvector(v1->x* n ,v1->y* n);
    return v3;
}

Pvector* Pvector::div(Pvector* v1,float n){
    Pvector *v3 = new Pvector(v1->x/ n ,v1->y/ n);
    return v3;
}

float Pvector::dot(Pvector* v1, Pvector* v2){
   return (v1->x*v2->x) + (v1->y*v2->y);
}

float Pvector::angleBetween(Pvector* v1, Pvector* v2)
{
   float theta = acosf(dot(v1,v2)/(v1->mag()*v2->mag()));
   return theta;
}

Pvector* Pvector::getNormalPoint(Pvector* p, Pvector* a, Pvector* b){
    Pvector *ap = sub(p,a);
    Pvector *ab = sub(b,a);

    ab->normalize();
    ab->mult(dot(ap,ab));

    Pvector* normalPoint = add(a,ab);

    return normalPoint;
}

float Pvector::dist(Pvector* a, Pvector* b){
    return sqrt(((a->x)-(b->x))*((a->x)-(b->x)) + ((a->y)-(b->y))*((a->y)-(b->y)));
}

File: UnitTesting/pvector.py
#Brief Rewritten and Refactored from code in C
#  - A toolbox that manipulate all the physics behind the simulation
# Detail
# - Pvector class contains all the necessary tools for doing maths
#   on a two-dimensional vector space. Most of the functions here
#   to generate ideal forces to drive the agents.
import math

class Pvector:

    def __init__(self, xCoordinate, yCoordinate):
        self.x = float(xCoordinate)
        self.y = float(yCoordinate)

    def add(self, pvector):
        x = self.x + pvector.x
        y = self.y + pvector.y
        return Pvector(x, y)

    def substract(self, pvector):
        x = self.x - pvector.x
        y = self.y - pvector.y
        return Pvector(x, y)

    def multiply(self, alpha):
        x = self.x * alpha
        y = self.y * alpha
        return Pvector(x, y)

    def divide(self, alpha):
        if alpha != 0:
            x = self.x/float(alpha)
            y = self.y/float(alpha)
            return Pvector(x, y)
        else:
            return Pvector(self.x, self.y)
            print('Division by zero')

    def addToSelf(self, pvector):
        self.x = self.x + pvector.x
        self.y = self.y + pvector.y

    def substractfromSelf(self, pvector):
        self.x = self.x - pvector.x
        self.y = self.y - pvector.y

    def multiplySelfByScalar(self, alpha):
        self.x = self.x * alpha
        self.y = self.y * alpha

    def divideSelfByScalar(self, alpha):
        if alpha != 0:
            self.x = self.x/float(alpha)
            self.y = self.y/float(alpha)
        else:
            print('Division by zero')

    def magnitudeSquared(self):
        return self.x*self.x + self.y*self.y

    def magnitude(self):
        return math.sqrt(self.x*self.x + self.y*self.y)

    def normalize(self):
        mag = self.magnitude()
        self.divideSelfByScalar(mag)

    def get(self):
        return self

    def setMagnitude(self, x):
        self.normalize()
        self.multiplySelfByScalar(x)

    def limitMagnitude(self, lim):
        if self.magnitude() > lim:
            self.setMagnitude(lim)

    def dotProduct(self, vector):
        x = self.x * vector.x
        y = self.y * vector.y
        return x + y

    def angleBetween(self, vector):
        angle = math.acosf(self.dotProduct(vector)/(self.magnitude() * vector.magnitude()))
        return angle

    def distanceBetween(self, vector):
        return math.sqrt((self.x - vector.x)*(self.x - vector.x) + (self.y - vector.y)*(self.y - vector.y))

    def getNormalPoint(self, b, p):
        pa = p.substract(self)
        ba = b.substract(self)
        ba.normalize()
        value = pa.dotProduct(ba)
        ba.multiplySelfByScalar(value)
        normalPoint = a.add(ba)
        return normalPoint
